src/uniqlist.c

```diff
@@ -228,3 +228,35 @@ SEXP nestedid(SEXP l, SEXP cols, SEXP order, SEXP grps, SEXP resetvals, SEXP mul
   UNPROTECT(1);
   return(ans);
 }
+
+SEXP uniqueNlogical(SEXP x, SEXP narmArg) {
+  // single pass; short-circuit and return as soon as all 3 values are found
+  if (!isLogical(x)) error("x is not a logical vector");
+  if (!isLogical(narmArg) || length(narmArg)!=1 || INTEGER(narmArg)[0]==NA_INTEGER) error("na.rm must be TRUE or FALSE");
+  bool narm = LOGICAL(narmArg)[0]==1;
+  int n = LENGTH(x);
+  if (n==0)
+    return( ScalarInteger(0) );  // empty vector
+  Rboolean first = INTEGER(x)[0];
+  int i=1;
+  while (i<n && INTEGER(x)[i]==first) i++;
+  if (i==n)
+    return( ScalarInteger( first==NA_INTEGER && narm ? 0 : 1) ); // all one value
+  Rboolean second = INTEGER(x)[i];
```

Comment:
@mattdowle if `na.rm`, we can short-circuit here if both `first` and `second are not `NA_INTEGER`

Reply:
it already does that, iiuc, on line 253 with the comment `// TRUE and FALSE found before any NA, but na.rm=TRUE so we're done`